### PR TITLE
Support both use_calc_stream and sync_op in send recv APIs

### PR DESCRIPTION
**Pull request by HermitSun**
### PR types

New features

### PR changes

APIs

### Describe

In the new communication library, we designed ProcessGroup to manage different communication group. Inside each process_group has its own stream which all communications in this group will be done on this stream. For high level API, like `distributed.all_reduce`, we use `use_calc_stream` to indicate whether this operation is sync or not. Notice that frequently add unnecessary cuda events may lead to low performance on some model. In order to achieve high performance, this pr add a new API name distributed.stream.all_reduce. This new API provided `use_calc_stream` and `sync_op` both.

- `sync_op`, indicate whether communication is sync or not.
- `use_calc_stream`, do communicate on calc stream, save the time of switching stream. Only work when `sync_op` is true.

### CODE REVIEW

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 1]
```diff
@@ -51,6 +51,17 @@ std::shared_ptr<ProcessGroupNCCL::NCCLTask> ProcessGroupNCCL::CreateTask(
       places, rank, comm_type, inputs);
 }
 
+std::shared_ptr<ProcessGroupNCCL::NCCLTask> ProcessGroupNCCL::CreateTask(
+    std::vector<Place> places,
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: const vector&

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 2]
```diff
@@ -406,6 +417,73 @@ void ProcessGroupNCCL::Collective(const phi::DenseTensor* in,
...
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (places_to_ncclcomm_.find(key) == places_to_ncclcomm_.end()) {
+      CreateNCCLManagerCache(key, places);
+    }
+  }
+
+  auto& nccl_comms = places_to_ncclcomm_[key];
+
+  SyncDefaultStream(places, places_to_events_[key], places_to_ctx_[key]);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: 这里可以根据use_calc_stream来判断需不需要同步计算流，Collective函数里也可以改一下

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 3]
```diff
@@ -406,6 +417,73 @@ void ProcessGroupNCCL::Collective(const phi::DenseTensor* in,
...
+                platform::DeviceContextPool::Instance().Get(places[i]))
+                ->stream();
+      } else {
+        nccl_stream = places_to_ctx_[key][i]->stream();
+      }
+      fn(tensors[i], nccl_comms[i]->GetNcclComm(), nccl_stream, dst_rank);
+    }
+  }
+
+  for (size_t i = 0; i < tensors.size(); ++i) {
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: 这里需要加判断，如果在计算流上不需要加event依赖

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 1 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 3 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 2 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 4]
```diff
@@ -264,10 +275,12 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupNCCL::Collective(
 
   auto& nccl_comms = places_to_ncclcomm_[key];
 
-  SyncDefaultStream(places, places_to_events_[key], places_to_ctx_[key]);
+  if (use_calc_stream) {
+    SyncDefaultStream(places, places_to_events_[key], places_to_ctx_[key]);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc:279]: 这个应该是不用计算流的时候，要让通信流等一下计算流，比如send一个tensor，这个tensor在计算流上计算出来的，为了保证通信流拿到的是正确的结果，需要加一个event。

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc [hunk 5]
```diff
@@ -671,6 +812,42 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupNCCL::Send_Partial(
...
+    int length,
+    bool sync_op,
+    bool use_calc_stream) {
+  phi::DenseTensor flatten_tensor;
+  flatten_tensor.ShareDataWith(tensors).Resize({tensors.numel()});
+
+  phi::DenseTensor shared_input = flatten_tensor.Slice(offset, offset + length);
+
+  std::vector<phi::DenseTensor> shared_tensors;
+  shared_tensors.push_back(shared_input);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]:
```suggestion
   std::vector<phi::DenseTensor> shared_tensors = {flatten_tensor.Slice(offset, offset + length)};
```

File: paddle/fluid/pybind/distributed_py.cc [hunk 6]
```diff
@@ -217,6 +234,30 @@ void BindDistributed(py::module *m) {
...
+                 int dst_rank,
+                 int nranks,
+                 int rank_id,
+                 bool sync_op) {
+                auto tensor = CastPyArg2Tensor(py_tensor.ptr(), 0);
+                auto dense =
+                    std::dynamic_pointer_cast<phi::DenseTensor>(tensor.impl());
+                int numel = (*dense).numel();
+                int send_numel = numel / nranks;
+                int offset = send_numel * rank_id;
```
Comment (LiYuRio) [inline paddle/fluid/pybind/distributed_py.cc:250]: 这个返回的应该是int64，大一点的tensor这就越界了，后面再统一改吧

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 4 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc:279]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 5 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.cc (hunk 2 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.cc]: done

**Review by LiYuRio — APPROVED** [review]

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM
单测LABEL设置

**Review by liuTINA0907 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**